PROCESSES.md

```diff
@@ -0,0 +1,147 @@
+Non-releasable code (tests, tools, analyzers) are not versioned. The
+precise meaning of a breaking change (dictating version number
+increments) is out of the scope of this document. (Jon is writing a
+blog post about this topic and will link to it when it is published.)
+
+The version number in the API catalog (and therefore in project
+files) is one of:
+
+- The mostly recently released package version
+- The version that's *about* to be released (see below)
+- A prerelease with a suffix of 00 to indicate that the *next* release
+  should be a first prerelease, typically of a new minor version,
+  due to changes that have been merged. This is to avoid accidentally
+  including new features in a patch release. For example, having
+  released Google.Cloud.Storage.V1 version 2.0.0, new features were
+  merged into the `master` branch and the version was immediately
+  changed to 2.1.0-alpha00
+
+Typically version numbers should be changed in a commit which does
+nothing else, for clarity. Include both the `apis.json` change and
+the project file changes that occur when the project generator has
+been run, in the same commit.
+
+Dependencies in the API catalog are specified as properties where
+the property name is the package name and the value is the version
+number. If the version number is left blank, then:
+
+- For GAX or gRPC dependencies, the version number is determined by
+  the project generator, to keep these dependencies in sync for all
+  packages.
+- For projects within the repository, a project reference is used.
+
+If a version number is specified, it's *always* a package reference.
+Importantly, project references should only be used between
+production packages when both packages will be published together.
+This avoids inconsistent builds where a package is built against a
+local version which is different from the declared dependency. *This
+isn't currently checked, but will be as part of the releasing
+process.*
+
+## Releasing
+
+Releasing consists of five steps:
+
+- Updating the version number in GitHub
+- Tagging the commit (typically the commit that updated the version
+  number)
+- Building and testing
+- Pushing the package to nuget.org
+- Updating the documentation in GitHub (in the `gh-pages` branch)
+
+### Current process
+
+- After the version number is committed, run `tagreleases.sh` from the
+  root directory to tag all updated versions.
+- Run `buildrelease.sh` manually, specifying one of the API names
+  and versions just tagged.
+- Go into the `releasebuild/nuget` directory and push
+- Make sure you have a clone of the gh-pages branch, copy the
+  documentation, commit and push.
+  
+Issues with this:
+
+- Integration tests are not run automatically
+- The build takes a long time as it rebuilds all packages, runs unit
+  tests for them all and builds all the documentation
+- There's no validation of project dependencies as described above
```

Comment:
I don't know how far we want to go in this testing - it blows up pretty quickly. Suppose we release a new version of Diagnostics which depends on Trace 1.0, which depends on GAX 2.0 which depends on Grpc.Core 1.4.  Now suppose there have since been versions Trace 1.1, GAX 2.1, Grpc.Core 1.5. How many permutations do we test?

- Trace 1.0, GAX 2.0, Grpc.Core 1.4
- Trace 1.0, GAX 2.0, Grpc.Core 1.5
- Trace 1.0, GAX 2.1, Grpc.Core 1.4
- Trace 1.0, GAX 2.1, Grpc.Core 1.5
- Trace 1.1, GAX 2.0, Grpc.Core 1.4
- Trace 1.1, GAX 2.0, Grpc.Core 1.5
- Trace 1.1, GAX 2.1, Grpc.Core 1.4
- Trace 1.1, GAX 2.1, Grpc.Core 1.5

It rapidly spirals out of control, IMO.

Perhaps we should run integration tests in two modes:

- Leave everything at minimally supported
- Upgrade everything to latest release

I think that would give a good balance of confidence vs reasonable speed. We *would* need to work out which "upstream" packages to run the integration tests on each time as well of course, which is interesting. Easier to do if we only consider packages within our own repo.

(What about for pre-releases? Do we upgrade everything to the latest pre-release? That may well not be valid...)

I suspect we need to talk about this separately...